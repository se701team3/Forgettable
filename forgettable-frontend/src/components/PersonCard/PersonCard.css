.MuiAvatarGroup-avatar {
    height: 20px;
    width: 20px;
    font-size: 12px;
    font-weight: bold;
}
