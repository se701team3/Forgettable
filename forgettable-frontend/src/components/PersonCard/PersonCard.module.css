@import '../../colours.css';

.PersonCard {
    width: 680px;
    height: 150px;
    border-radius: 13px;
    background-color: var(--card);
    transition: 0.3s;
    display: flex;
    padding: 28px;
    box-sizing: border-box;
}

.ContentContainer {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.PersonCard:hover {
    background-color: var(--hilit);
}

.InformationContainer {
    height: 100%;
    flex: 1;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: stretch;
}

.MainInformationContainer {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.IdentityInformation {
    height: 100%;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    flex: 1;
}

.IdentityInformation > h2 {
    font-size: 25px;
    font-weight: 600;
    /* line-height: 0; */
    margin: 0
}

.IdentityInformation > p {
    font-size: 16px;
    /* line-height: 0; */
    margin: 2px 0;
    color: var(--ltxt2)
}

.SupplementaryInformationContainer {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.SupplementaryInformationContainer > p {
    margin: 0;
    font-size: 14px;
    color: var(--ltxt2)
}

.SupplementaryInformationContainer > div {
    width: 50px;
    height: 20px;
    display: flex;
    justify-content: center;
}

.IconButton {
    align-self: flex-start;
}

