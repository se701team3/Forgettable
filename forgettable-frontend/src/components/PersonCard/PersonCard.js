import {Avatar, AvatarGroup, StyledEngineProvider, SvgIcon} from '@mui/material';
import React from 'react';
import classes from './PersonCard.module.css';
import IconButton from '@mui/material/IconButton';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import MoreHorizIcon from '@mui/icons-material/MoreHoriz';
import './PersonCard.css';
import {convertSocialMediaToIcon} from '../../functions/socialMediaIconConverter';
import PropTypes from 'prop-types';
import {getFirstMetTimeString, getDateLastMetString} from '../../functions/dateFormatter';

/*
 * Component for displaying information of a person in a list.
 * This is the full version of the person card.
 *
 * Author: Mercury Lin (lin8231)
 */
const PersonCard = (props) => {
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const socialMediaIcons = props.socialMedias &&
    props.socialMedias.map((socialMedia) => {
      return (
        <Avatar
          key={socialMedia}
          src={convertSocialMediaToIcon(socialMedia)}
          alt={socialMedia}
        >
        </Avatar>
      );
    });

  const menuItems = [
    {
      label: 'Edit',
      action: props.onEdit,
    },
    {
      label: 'Delete',
      action: props.onDelete,
    },
  ];

  return (
    <StyledEngineProvider injectFirst>
      <div className={classes.PersonCard}>
        <div className={classes.ContentContainer}>
          <Avatar
            alt={props.name}
            // this style is written inline because MUI does not support className
            style={{
              height: '90px',
              width: '90px',
              marginRight: '28px',
              backgroundColor:
                getComputedStyle(document.body).getPropertyValue('--prmry'),
              fontSize:
                getComputedStyle(document.body).getPropertyValue('--text-xxlarge')
              ,
            }}
            src={props.img}
          />
          <div className={classes.InformationContainer}>
            <div className={classes.MainInformationContainer}>
              <div className={classes.IdentityInformation}>
                <h2 data-testid="name-element">{props.name}</h2>
                <p data-testid="first-met-element">
                  {'First met ' + getFirstMetTimeString(props.firstMet)}
                </p>
              </div>
              <IconButton
                className={classes.IconButton}
                aria-label="more"
                id="long-button"
                aria-controls={open ? 'long-menu' : undefined}
                aria-expanded={open ? 'true' : undefined}
                aria-haspopup="true"
                onClick={handleClick}
              >
                <MoreHorizIcon />
              </IconButton>
              <Menu
                id="long-menu"
                MenuListProps={{
                  'aria-labelledby': 'long-button',
                  'disablePadding': true,
                }}
                anchorEl={anchorEl}
                open={open}
                onClose={handleClose}
                PaperProps={{
                // this style is written inline because MUI does not support className
                  style: {
                    maxHeight: 35 * 4.5,
                    width: '130px',
                    boxShadow: 'none',
                    outline: '1px solid black',
                  },
                }}
                anchorOrigin={{
                  vertical: 'bottom',
                  horizontal: 'center',
                }}
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
              >
                {menuItems.map((menuItem) => {
                  return (
                    <MenuItem
                      divider={true}
                      key={menuItem.label}
                      onClick={menuItem.action}
                    >
                      {menuItem.label}
                    </MenuItem>
                  );
                },
                )}
              </Menu>
            </div>
            <div className={classes.SupplementaryInformationContainer}>
              <p className={classes.Encounters}
                data-testid="encounters-element"
              >
                  Encounters: {props.numEncounters} times
              </p>
              <p className={classes.LastMet}
                data-testid="last-met-element"
              >
                Date last met: {getDateLastMetString(props.lastMet)}
              </p>
              <div className={classes.SocialMediaContainer}>
                <AvatarGroup max={2}
                  spacing={2}
                  aria-label="social-media-icons"
                  data-testid="social-media-icons-element"
                >
                  { socialMediaIcons }
                </AvatarGroup>
              </div>
            </div>
          </div>
        </div>
      </div>
    </StyledEngineProvider>
  );
};

PersonCard.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  img: PropTypes.string,
  socialMedias: PropTypes.arrayOf(PropTypes.string),
  numEncounters: PropTypes.number,
  lastMet: PropTypes.instanceOf(Date),
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  firstMet: PropTypes.instanceOf(Date),
};


export default PersonCard;
