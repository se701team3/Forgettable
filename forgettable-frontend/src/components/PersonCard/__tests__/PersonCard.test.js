import TestRenderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import PersonCard from '../PersonCard';
import React from 'react';
import {render, fireEvent, cleanup} from '@testing-library/react';
import {
  within,
  getByText,
  getByTestId,
} from '@testing-library/dom';

it('renders PersonCard UI with correct hierarchy', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<PersonCard
    name={'Mercury Lin'}
    onDelete={jest.fn()}
    onEdit={jest.fn()}
  />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});

afterEach(cleanup);

it('Name prop passed in is displayed correctly', () => {
  const name = 'Mercury Lin';

  const {getByTestId} = render(
      <PersonCard
        name={name}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
      />);

  const {getByText} = within(getByTestId('name-element'));

  expect(getByText(name)).toBeInTheDocument();
});

it('renders the correct number of social media icons', () => {
  const socialMedias = ['instagram', 'facebook'];

  const {getByTestId} = render(
      <PersonCard
        onDelete={jest.fn()}
        onEdit={jest.fn()}
        socialMedias={socialMedias}
        name={'Mercury Lin'}
      />);

  const socialMediaContainer = getByTestId('social-media-icons-element');

  expect(socialMediaContainer.childElementCount).toBe(socialMedias.length);
});

it('displays the correct number of encounters', () => {
  const numEncounters = 14;

  const {getByTestId} = render(
      <PersonCard
        name={'Mercury Lin'}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
        numEncounters={numEncounters}
      />);

  const {getByText} = within(getByTestId('encounters-element'));

  expect(
      getByText('Encounters: ' + numEncounters + ' times'),
  ).toBeInTheDocument();
});

it('displays the correct format for lastMet', () => {
  const lastMet = new Date(1646965503063);

  const {getByTestId} = render(
      <PersonCard
        name={'Mercury Lin'}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
        lastMet={lastMet}
      />);

  const {getByText} = within(getByTestId('last-met-element'));

  expect(
      getByText('Date last met: 11/03/2022'),
  ).toBeInTheDocument();
});

it('displays the correct format for firstMet for an existing date', () => {
  const twoYearsAgo = new Date(Date.now() - (365 * 24 * 60 * 60 * 1000 * 2));

  const {getByTestId} = render(
      <PersonCard
        name={'Mercury Lin'}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
        firstMet={twoYearsAgo}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met 2 years ago'),
  ).toBeInTheDocument();
});

it('displays the firstMet string when there is no date', () => {
  const {getByTestId} = render(
      <PersonCard
        name={'Mercury Lin'}
        onDelete={jest.fn()}
        onEdit={jest.fn()}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met once upon a time'),
  ).toBeInTheDocument();
});
