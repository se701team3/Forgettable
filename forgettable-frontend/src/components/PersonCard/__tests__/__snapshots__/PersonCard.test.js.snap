// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PersonCard UI with correct hierarchy 1`] = `
<StyledEngineProvider
  injectFirst={true}
>
  <div
    className="PersonCard"
  >
    <div
      className="ContentContainer"
    >
      <ForwardRef(Avatar)
        alt="Mercury Lin"
        style={
          Object {
            "backgroundColor": "",
            "fontSize": "",
            "height": "90px",
            "marginRight": "28px",
            "width": "90px",
          }
        }
      />
      <div
        className="InformationContainer"
      >
        <div
          className="MainInformationContainer"
        >
          <div
            className="IdentityInformation"
          >
            <h2
              data-testid="name-element"
            >
              Mercury Lin
            </h2>
            <p
              data-testid="first-met-element"
            >
              First met once upon a time
            </p>
          </div>
          <ForwardRef(IconButton)
            aria-haspopup="true"
            aria-label="more"
            className="IconButton"
            id="long-button"
            onClick={[Function]}
          >
            <Memo />
          </ForwardRef(IconButton)>
          <ForwardRef(Menu)
            MenuListProps={
              Object {
                "aria-labelledby": "long-button",
                "disablePadding": true,
              }
            }
            PaperProps={
              Object {
                "style": Object {
                  "boxShadow": "none",
                  "maxHeight": 157.5,
                  "outline": "1px solid black",
                  "width": "130px",
                },
              }
            }
            anchorEl={null}
            anchorOrigin={
              Object {
                "horizontal": "center",
                "vertical": "bottom",
              }
            }
            id="long-menu"
            onClose={[Function]}
            open={false}
            transformOrigin={
              Object {
                "horizontal": "right",
                "vertical": "top",
              }
            }
          >
            <ForwardRef(MenuItem)
              divider={true}
              onClick={[MockFunction]}
            >
              Edit
            </ForwardRef(MenuItem)>
            <ForwardRef(MenuItem)
              divider={true}
              onClick={[MockFunction]}
            >
              Delete
            </ForwardRef(MenuItem)>
          </ForwardRef(Menu)>
        </div>
        <div
          className="SupplementaryInformationContainer"
        >
          <p
            className="Encounters"
            data-testid="encounters-element"
          >
            Encounters: 
            0
             times
          </p>
          <p
            className="LastMet"
            data-testid="last-met-element"
          >
            Date last met: 
            never :(
          </p>
          <div
            className="SocialMediaContainer"
          >
            <ForwardRef(AvatarGroup)
              aria-label="social-media-icons"
              data-testid="social-media-icons-element"
              max={2}
              spacing={2}
              sx={
                Object {
                  "& .MuiAvatar-root": Object {
                    "fontSize": 15,
                    "height": 20,
                    "width": 20,
                  },
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</StyledEngineProvider>
`;
